api/document/serializers/doc_cursor.py

```diff
@@ -118,15 +118,55 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.context['parent_serializer'] = self
 
+    @property
+    def is_root(self) -> bool:
+        return self.context.get('is_root', True)
+
     def get_meta(self, instance):
         """Include meta data which applies to the whole node."""
         meta = Meta(
             instance,
-            self.context.get('is_root', True),
+            self.is_root,
             self.context.get('policy'),
         )
         return MetaSerializer(meta, context={'parent_serializer': self}).data
 
+    def validate_footnote_citations(self, data: PrimitiveDict,
+                                    footnote_emblems: Set[str]):
+        citations = [
+            util.get_content_text(f['inlines']).strip()
+            for f in util.iter_inlines(data['content'])
+            if f['content_type'] == 'footnote_citation'
+        ]
```

Comment:
Inspecting the `PrimitiveDict` this deeply feels wrong. We have a somewhat sophisticated set of navigation tools for DocCursor already -- should we delay this validation until we have more of the tree constructed?

Reply:
So I felt that tension while writing this code, but there are a number of different competing issues at play:

1. DRF's serialization framework _really_ wants validation to occur on data structures that are completely independent of the database, whereas `DocCursor` is fairly coupled to the database. So using `DocCursor` during validation feels like fighting against DRF, at least a little bit.

2. `DocCursor` by itself doesn't have any way of representing annotations/inlines as part of a tree;  though the `JSONAwareCursor` does, its accesses them as primitive dicts, so we'd still need to write our own helpers to navigate through those, like `get_content_text()` and `iter_inlines()`.  So it feels like extra cognitive weight to treat _one_ aspect of the content (the nodes) as a nice `DiGraph`-based tree, while treating another aspect of the content (the content/inlines) as not-a-tree, even though it ultimately is, at least the way the client has delivered it to us.

That said, I could see a middle ground where we at least start constructing a `DocCursor`-based tree, independent of the database, in the `to_internal_value()` methods of our serializers, though we'd need to either come up with an alternative `DiGraph` structure for our annotations or keep handling them as JSON blobs accessible via a `JSONAwareCursor`.  But that said, it feels both like a non-trivial refactoring, and like something we could do later on, independently of this PR (though I suppose this PR does slightly increase the amount of code we'd need to migrate over to the `DocCursor`-based way of doing things).

At the very least, could I commit the _tests_ from this PR as `xfail` tests, so that we could fix them later, through whatever means we eventually decide on?

Reply:
You bring up very good points. I _think_ DRF may be the wrong layer for this sort of validation. I wonder if there's an analogy why Django's models -- fields have validations, models can have validations cross field, but cross-model validations are in the form-handling layer.

Regarding annotation tracking, let me take a crack at #823; I think that might be a cleaner way to keep all the models in memory _and_ keep them related to each other.

All that said, this is definitely an improvement over _no_ validations. I just wanted to pause us a little to think through some alternatives.

Reply:
Noice, that sounds good!

How soon are you thinking of forging #823?  I had an idea for unifying `json_importer.annotations` with `serializers.content`, to reduce the number of places new annotation types had to create new code at, but it can wait...